gwcs/converters/wcs.py

```diff
@@ -0,0 +1,201 @@
+    tags = ["tag:stsci.edu:gwcs/wcs-*"]
+    types = ["gwcs.wcs.WCS"]
+
+    def from_yaml_tree(self, node, tag, ctx):
+        from ..wcs import WCS
+        return WCS(node['steps'], name=node['name'])
+
+    def to_yaml_tree(self, gwcsobj, tag, ctx):
+        return {
+            'name': gwcsobj.name,
+            'steps': gwcsobj.pipeline
+        }
+
+
+class StepConverter(Converter):
+    tags = ["tag:stsci.edu:gwcs/step-*"]
+    types = ["gwcs.wcs.Step"]
+
+    def from_yaml_tree(self, node, tag, ctx):
+        from ..wcs import Step
+        return Step(frame=node['frame'], transform=node.get('transform', None))
+
+    def to_yaml_tree(self, step, tag, ctx):
+        return {
+            'frame': step.frame,
+            'transform': step.transform
+        }
+
+
+class FrameConverter(Converter):
+    tags = ["tag:stsci.edu:gwcs/frame-*"]
+    types = ["gwcs.coordinate_frames.CoordinateFrame"]
+
+    def _from_yaml_tree(self, node, tag, ctx):
+        kwargs = {'name': node['name']}
+
+        if 'axes_type' in node and 'naxes' in node:
+            kwargs.update({
+                'axes_type': node['axes_type'],
+                'naxes': node['naxes']})
+
+        if 'axes_names' in node:
+            kwargs['axes_names'] = node['axes_names']
+
+        if 'reference_frame' in node:
+            kwargs['reference_frame'] = node['reference_frame']
+
+        if 'axes_order' in node:
+            kwargs['axes_order'] = tuple(node['axes_order'])
+
+        if 'unit' in node:
+            kwargs['unit'] = tuple(node['unit'])
+
+        if 'axis_physical_types' in node:
+            kwargs['axis_physical_types'] = tuple(node['axis_physical_types'])
+
+        return kwargs
+
+    def _to_yaml_tree(self, frame, tag, ctx):
+        from ..coordinate_frames import CoordinateFrame
+
+        node = {}
+
+        node['name'] = frame.name
+
+        # We want to check that it is exactly this type and not a subclass
+        if type(frame) is CoordinateFrame:
+            node['axes_type'] = frame.axes_type
+            node['naxes'] = frame.naxes
+
+        if frame.axes_order is not None:
+            node['axes_order'] = list(frame.axes_order)
+
+        if frame.axes_names is not None:
+            node['axes_names'] = list(frame.axes_names)
+
+        if frame.reference_frame is not None:
+            node['reference_frame'] = frame.reference_frame
+
+        if frame.unit is not None:
+            node['unit'] = list(frame.unit)
+
+        if frame.axis_physical_types is not None:
+            node['axis_physical_types'] = list(frame.axis_physical_types)
+
+        return node
+
+    def from_yaml_tree(self, node, tag, ctx):
+        from ..coordinate_frames import CoordinateFrame
+        node = self._from_yaml_tree(node, tag, ctx)
+        return CoordinateFrame(**node)
+
+    def to_yaml_tree(self, frame, tag, ctx):
+        return self._to_yaml_tree(frame, tag, ctx)
+
+
+class Frame2DConverter(FrameConverter):
```

Comment:
A tactic that we use in asdf-astropy to cut down on duplicate code is to initialize the converter class with tag and class name arguments:

https://github.com/astropy/asdf-astropy/blob/main/asdf_astropy/converters/transform/core.py#L225

We then instantiate that class several times when defining the extension:

https://github.com/astropy/asdf-astropy/blob/main/asdf_astropy/extensions.py#L41-L52

Reply:
@eslavich if I implement this suggestion, I get the following failure(s) in unit tests:

```python
AttributeError: 'Frame2D' object has no attribute 'param_names'
```

Reply:
Hmm... well I think we should leave it be for now, and maybe try to reuse a class later.